src/librustc_mir/transform/const_prop.rs

```diff
@@ -272,78 +277,71 @@ fn type_size_of<'a, 'tcx>(tcx: TyCtxt<'a, 'tcx, 'tcx>,
+    ) -> IndexVec<Local, bool> {
         let mut cpv = CanConstProp {
             can_const_prop: IndexVec::from_elem(true, &mir.local_decls),
             found_assignment: IndexVec::from_elem(false, &mir.local_decls),
+            mir,
+            tcx,
+            param_env,
         };
         for (local, val) in cpv.can_const_prop.iter_enumerated_mut() {
-            *val = mir.local_kind(local) == LocalKind::Temp;
+            *val = mir.local_kind(local) != LocalKind::Arg;
         }
         cpv.visit_mir(mir);
         cpv.can_const_prop
     }
 }
 
-fn place_to_local(mut place: &Place) -> Option<Local> {
-    while let Place::Projection(ref proj) = place {
-        place = &proj.base;
-    }
-    if let Place::Local(local) = *place {
-        Some(local)
-    } else {
-        None
-    }
-}
-
-impl<'tcx> Visitor<'tcx> for CanConstProp {
-    fn visit_statement(
+impl<'a, 'b, 'tcx> Visitor<'tcx> for CanConstProp<'a, 'b, 'tcx> {
+    fn visit_local(
         &mut self,
-        block: BasicBlock,
-        statement: &Statement<'tcx>,
-        location: Location,
+        &local: &Local,
+        context: PlaceContext<'tcx>,
+        _: Location,
     ) {
-        self.super_statement(block, statement, location);
-        match statement.kind {
-            StatementKind::SetDiscriminant { ref place, .. } |
-            StatementKind::Assign(ref place, _) => {
-                if let Some(local) = place_to_local(place) {
-                    if self.found_assignment[local] {
-                        self.can_const_prop[local] = false;
-                    } else {
-                        self.found_assignment[local] = true
-                    }
-                }
+        use rustc::mir::visit::PlaceContext::*;
+        match context {
+            // Constants must have at most one write
+            // FIXME(oli-obk): we could be more powerful here, if the multiple writes
+            // only occur in independent execution paths
+            Store => if self.found_assignment[local] {
+                self.can_const_prop[local] = false;
+            } else {
+                self.found_assignment[local] = true
             },
-            StatementKind::InlineAsm { ref outputs, .. } => {
-                for place in outputs {
-                    if let Some(local) = place_to_local(place) {
-                        if self.found_assignment[local] {
-                            self.can_const_prop[local] = false;
-                        } else {
-                            self.found_assignment[local] = true
-                        }
-                        return;
-                    }
+            // Reading constants is allowed an arbitrary number of times
+            Copy | Move |
+            StorageDead | StorageLive |
+            Validate |
+            Inspect => {},
+            Borrow { kind: BorrowKind::Shared, .. } => {
+                // cannot const prop immutable borrows of types with interior mutability
```

Comment:
Do we need this special-case?

Reply:
we can just forbid all borrows?

Reply:
Yeah, only named variables can be both used directly and borrowed, so this should be fine, I think.